### DISCUSSION TITLE
Role Based Access control of Data

### DISCUSSION THREAD

**Discussion by t2the2ndpower** [Budibase/budibase#965] [Help] [answered]
Hi there,

Some how I found Budibase while looking for the most fitting lo/no code builder for all of my wild hair ideas (and I'm glad I did!).   And have been playing around with it for the last week.

I wanted to know if there was a suggested method for tying data to the user that created it (i.e. if the user created it, the user can see it. If the user did NOT create it, that user can NOT see it).  Full disclosure, I'm a bit of a noob (thus the love of lo/no code environments) so if this is already written up somewhere (and I just didn't see it) please shoot me the link.  I've gone through the videos and tried to sift through the repo (which is kind of like Spanglish to me) and hadn't come across anything that sounded like what I was looking for.

Thanks for any tips you may have and may the Budi be with you!

**Comment by shogunpurple** [comment, marked as answer]
Hi @t2the2ndpower! Thanks for taking the time to open a discussion.

The platform does not currently track which "builder user" created the data. The concept of keeping a history of edits/audit trail is something that we will implement at some point, but is not on our immediate roadmap. 

Fortunately there are _some_ ways that you could solve this problem with the current platform. I've split them into what you can do right now and what you will be able to do in the very near future.

### What you can do now
It doesn't satisfy your use case exactly, but you can restrict access to certain screens **by role** making use of the role based access system to create a screen specifically for a role. Basically this requires the following:
- Create a table or view you only want specific users to see
- Create a role (or use an existing one) that users must have to see your data
- Go to design panel and create a screen with that role
- Add components etc to that screen to display your data
- When users of the role log in, they will be able to see that screen. Other users will not

Here are some videos that show you how to do that.

**Create a new role and a new screen for that role**
https://www.loom.com/share/1a332cfe924c471988c31d24d95feceb

**Only show secret data when the locked user logs in**
https://www.loom.com/share/232ba2ff97244d65b2f1c632263e26ff

### Roadmap - What you will soon be able to do
You will be able to assign roles to a specific table, preventing read and write access to them without having to show/hide them with different screens as above. This hasn't been started yet but will be done in the coming weeks. (https://github.com/Budibase/budibase/issues/915)

Alternatively, you will also be able to show only data from a relationship associated with a user. For example:
- I create a "tasks" table
- I create a relationship between the users table and tasks
- In the frontend, I can show a grid with "The current users tasks". Only the logged in user will be able to see their own data

This is actually done already, but not yet released. It will be available in the product by the end of the week.

**Reply by t2the2ndpower** [reply]
Thanks so much for awesome detailed response!  The videos are super helpful. Being able to restrict screens by roles will help with my 'User Admin' area (for an 'Agency' like experience). 

I'm very much excited to see what you have planned for the upcoming release as it sounds like it will actually get me exactly what I'm looking for... "Only the logged in user will be able to see their own data."  Whoo Hoo!

I hope to have a beta version of my little app or something worth sharing in a few weeks.

Thanks again and I'm really liking Budibase. Kudos to you clever gents!

**Reply by shogunpurple** [reply]
No problem at all. Glad we could help! Don’t hesitate to let us know if you have any more questions.

We are excited too - watch this space :)

**Comment by unknown1337** [comment]
I am wondering if there already is any update on a part of the initial question:
- will there be a feature for the **backend** to only allow users to update the table rows/ entries that are owned/ created/ linked by them?

Background: this would allow the usage of budibase for a much broader audience in stead of an internal tool only. In case e.g. a user creates an account and then is able to add rows to a table of e.g. owned cars, he should only have rights to get a list and change the cars that **he owns**. 

Current situation: 
1. several access roles
2. read access to a table (in my example BudibaseDB which e.g. contains personal invoices/ todo's/..) can be limited to a access role (e.g. basic) (backend secured I assume, not frontend) 
3. in case all users belong to e.g. the "basic" access role, they all could (if they change the frontend javascript) access the data of all users that is stored in the same table which is **NOK** for many applications

Desired situation (for me and possibly many others):
1. users can create an account by filling in an form (+confirmation link)
2. users can write data to a given table 
3. users can read data from a given table IFF they have created the original row/ are linked in that row/entry

Thanks!!!

**Reply by mjashanks** [reply]
Thanks for writing this up - it will be super useful when we come to implement this. 

However, I don't think we will be doing this one for a while. I do agree that it would be a fantastic feature but there is a lot of complexity in it. Not on the immediate roadmap - although if lots of people weigh in here, that could change!
